File: src/LookingGlass.View/RegionTracker.cs
/* Copyright (c) 2008 Robert Adams
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * The name of the copyright holder may not be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
using System;
using System.Collections.Generic;
using System.Text;
using LookingGlass.Framework;
using LookingGlass.Framework.Logging;
using LookingGlass.Framework.Modules;
using LookingGlass.Framework.Parameters;
using LookingGlass.Rest;
using LookingGlass.Renderer;
using LookingGlass.World;
using OMV = OpenMetaverse;
using OMVSD = OpenMetaverse.StructuredData;

namespace LookingGlass.View {
    /// <summary>
    /// Watch the comings and goings of the regions and handle the level of detail
    /// that the regions are displayed in.
    /// </summary>
public class RegionTracker : IRegionTrackerProvider, IModule {

protected RestHandler m_regionRestHandler;

protected IWorld m_world;
protected IRenderProvider m_renderer;

#region IMODULE
    protected string m_moduleName;
    public string ModuleName { get { return m_moduleName; } set { m_moduleName = value; } }

    protected LookingGlassBase m_lgb = null;
    public LookingGlassBase LGB { get { return m_lgb; } }

    public IAppParameters ModuleParams { get { return m_lgb.AppParams; } }

    public RegionTracker() {
        // default to the class name. The module code can set it to something else later.
        m_moduleName = System.Reflection.MethodBase.GetCurrentMethod().DeclaringType.Name;
    }

    // IModule.OnLoad
    public virtual void OnLoad(string modName, LookingGlassBase lgbase) {
        LogManager.Log.Log(LogLevel.DINIT, "RegionTracker.OnLoad()");
        m_moduleName = modName;
        m_lgb = lgbase;
        // set the parameter defaults
        ModuleParams.AddDefaultParameter(ModuleName + ".Regions.Enable", "true",
                    "Whether to make region information available");
        ModuleParams.AddDefaultParameter(ModuleName + ".World.Name", "World",
                    "Name of world module to track entities in");
        ModuleParams.AddDefaultParameter(ModuleName + ".Renderer.Name", "Renderer",
                    "Name of renderer module for display of region details");
    }
   
    // IModule.AfterAllModulesLoaded
    public virtual bool AfterAllModulesLoaded() {
        LogManager.Log.Log(LogLevel.DINIT, "EntityTracker.AfterAllModulesLoaded()");
        // connect to the world and listen for entity events
        m_world = (IWorld)LGB.ModManager.Module(ModuleParams.ParamString(ModuleName + ".World.Name"));
        m_renderer = (IRenderProvider)LGB.ModManager.Module(ModuleParams.ParamString(ModuleName + ".Renderer.Name"));
        if (ModuleParams.ParamBool(ModuleName + ".Regions.Enable")) {
            m_world.OnWorldRegionNew += new WorldRegionNewCallback(World_OneWorldRegionNew);
            m_world.OnWorldRegionRemoved += new WorldRegionRemovedCallback(World_OneWorldRegionRemoved);
            m_world.OnWorldRegionUpdated += new WorldRegionUpdatedCallback(World_OneWorldRegionUpdated);
        }

        if (ModuleParams.ParamBool(ModuleName + ".Regions.Enable")) {
            m_regionRestHandler = new RestHandler("/Tracker/Regions/", new RegionInformation(this));
        }
        return true;
    }

    // IModule.Start
    public virtual void Start() {
        return;
    }

    // IModule.Stop
    public virtual void Stop() {
        return;
    }

    // IModule.PrepareForUnload
    public virtual bool PrepareForUnload() {
        if (ModuleParams.ParamBool(ModuleName + ".Regions.Enable")) {
            m_world.OnWorldRegionNew -= new WorldRegionNewCallback(World_OneWorldRegionNew);
            m_world.OnWorldRegionRemoved -= new WorldRegionRemovedCallback(World_OneWorldRegionRemoved);
            m_world.OnWorldRegionUpdated -= new WorldRegionUpdatedCallback(World_OneWorldRegionUpdated);
        }
        return false;
    }
#endregion IMODULE

#region EVENT PROCESSING
    void World_OneWorldRegionNew(RegionContextBase rcontext) {
        // for the moment, any close by region is good enough for focus
        if (m_renderer != null) {
            LogManager.Log.Log(LogLevel.DWORLDDETAIL, "RegionTracker: setting focus region {0}", rcontext.Name);
            m_renderer.SetFocusRegion(rcontext);
        }
    }
    void World_OneWorldRegionRemoved(RegionContextBase rcontext) {
    }
    void World_OneWorldRegionUpdated(RegionContextBase rcontext, UpdateCodes what) {
    }
#endregion EVENT PROCESSING

#region RESPONSE DATA CONTSRUCTION
    private class RegionInformation : IDisplayable {
        RegionTracker m_tracker;
        public RegionInformation(RegionTracker regTrack) {
            m_tracker = regTrack;
        }
        public OMVSD.OSDMap GetDisplayable() {
            return new OMVSD.OSDMap();
        }
    }
#endregion RESPONSE DATA CONSTRUCTION

}
}

File: src/LookingGlassOgre/RegionTracker.cpp
/* Copyright (c) Robert Adams
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * The name of the copyright holder may not be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include "StdAfx.h"
#include "RendererOgre.h"
#include "Region.h"
#include "RegionTracker.h"

namespace LG {

RegionTracker* RegionTracker::m_instance = NULL;

RegionTracker::RegionTracker() {
	m_focusRegion = NULL;
}
RegionTracker::~RegionTracker() {
}

// Add a region to the regions being tracked.
// This creates the tracking Region instance, creates a scene node for the region
// and adds scene nodes to the region for the terrain and water.
// BETWEEN FRAME OPERATION
void RegionTracker::AddRegion(const char* regionSceneName,
							  double globalX, double globalY, double globalZ,
							  const float sizeX, const float sizeY, const float waterHeight) {
	Ogre::String regionSceneNodeName = Ogre::String(regionSceneName);
	// if already have defn for region, return
	LG::Log("RegionTracker::AddRegion: adding region %s", regionSceneName);
	Region* regn = FindRegion(regionSceneNodeName);
	if (regn != NULL) {
		return;
	}
	// create Region class
	regn = new Region();
	regn->Name = regionSceneNodeName;
	m_regions.insert(std::pair<Ogre::String, Region*>(regionSceneNodeName, regn));
	regn->Init(globalX, globalY, globalZ, sizeX, sizeY, waterHeight);
	RecalculateLocalCoords();
}

Region* RegionTracker::FindRegion(Ogre::String nam) {
	RegionHashMap::iterator intr = m_regions.find(nam);
	if (intr == m_regions.end()) {
		return NULL;
	}
	return intr->second;
}

std::list<Region*> RegionTracker::GetRegions() {
	std::list<Region*> regns;
	for (RegionHashMap::iterator intr = m_regions.begin(); intr != m_regions.end(); intr++) {
		Region* otherRegn = intr->second;
		regns.push_back(otherRegn);
	}
	return regns;
}

void RegionTracker::UpdateTerrain(const char* regnName, const int width, const int length, const float* hm) {
	Ogre::String regionName = Ogre::String(regnName);
	Region* regn = FindRegion(regionName);
	if (regn != NULL) {
		regn->UpdateTerrain(width, length, hm);
	}
	else {
		LG::Log("RegionTracker::UpdateTerrain: region not found so terrain not updated: %s", regnName);
	}
}

// Set a focus region and update all region's local coords relative to the focus region
// BETWEEN FRAME OPERATION
void RegionTracker::SetFocusRegion(Ogre::String regionName) {
	LG::Log("RegionTracker::SetFocusRegion: %s", regionName.c_str());
	for (RegionHashMap::iterator intr = m_regions.begin(); intr != m_regions.end(); intr++) {
		Region* otherRegn = intr->second;
		otherRegn->SetFocusRegion(false);
	}
	Region* regn = FindRegion(regionName);
	if (regn != NULL) {
		m_focusRegion = regn;
		regn->SetFocusRegion(true);
		RecalculateLocalCoords();
	}
}

// recalcualate the local coords based on the focus region
// BETWEEN FRAME OPERATION
void RegionTracker::RecalculateLocalCoords() {
	Region* regn = GetFocusRegion();
	if (regn != NULL) {
		for (RegionHashMap::iterator intr = m_regions.begin(); intr != m_regions.end(); intr++) {
			Region* otherRegn = intr->second;
			otherRegn->CalculateLocal(regn->GlobalX, regn->GlobalY, regn->GlobalZ);
		}
	}
	// Do I have to update all the scene nodes to tell them that stuff has changed?
	LG::RendererOgre::Instance()->m_sceneMgr->getRootSceneNode()->needUpdate();
}

// Focusing a region causes the coordinate system to be zeroed to that point. Since the camera
// operates in global coordinates, it must be offset for the focus region.
Ogre::Vector3 RegionTracker::PositionForFocusRegion(Ogre::Vector3 pos) {
	return PositionCameraForFocusRegion(pos.x, pos.y, pos.z);
}

// Focusing a region causes the coordinate system to be zeroed to that point. Since the camera
// operates in global coordinates, it must be offset for the focus region.
Ogre::Vector3 RegionTracker::PositionCameraForFocusRegion(double px, double py, double pz) {
	Region* fRegion = GetFocusRegion();
	Ogre::Vector3 newPos;
	if (fRegion != NULL) {
		newPos.x = (Ogre::Real)(px - fRegion->GlobalX);
		newPos.y = (Ogre::Real)(py - fRegion->GlobalY);
		newPos.z = (Ogre::Real)(pz - fRegion->GlobalZ);
	}
	else {
		// if no focus region, just return what they passed
		newPos.x = (Ogre::Real)px;
		newPos.y = (Ogre::Real)py;
		newPos.z = (Ogre::Real)pz;
	}
	// cam->setPosition(px - fRegion->GlobalX, py - fRegion->GlobalY, pz - fRegion->GlobalZ);
	return newPos;
}

// return focus region or NULL if none
Region* RegionTracker::GetFocusRegion() {
	return m_focusRegion;
}

void RegionTracker::SetRegionDetail(Ogre::String regionName, const RegionRezCode LODLevel) {
	Region* regn = FindRegion(regionName);
	if (regn != NULL) {
		regn->ChangeRez(LODLevel);
	}
}

}


